lib/puppet/util/execution.rb

```diff
@@ -89,6 +89,47 @@ def self.execpipe(command, failonfail = true)
     output
   end
 
+  # Wait for a process to exit or kill it after a given timeout
+  #
+  # This function polls for the exit status of a child process. If the
+  # supplied timeout is exceeded, then the child is terminated.
+  #
+  # @param child_pid [Integer] The PID of the process to monitor.
+  # @param timeout [Integer] The maximum amount of time to wait, in seconds.
+  #
+  # @return [Array<Integer, Process::Status>] A two element array containing
+  #   the PID of the monitored process along with a Process::Status object
+  #   containing its exit status as returned by Process.waitpid2.
+  def self.wait_with_timeout(child_pid, timeout)
+    # TODO: Use a monotonic clock.
+    deadline = Time.now + timeout
+    # The Process module on Windows only supports :KILL, unless a win32
+    # extension gem is used. Even then, the :TERM supplied is equivalent
+    # to :KILL
+    kill_signal = Puppet.features.microsoft_windows? ? :KILL : :TERM
+
+    loop do
+      # Poll status of child process without blocking. Returns nil
+      # if the process has not exited.
+      status = Process.waitpid2(child_pid, Process::WNOHANG)
```

Comment:
I'm not sure if `waitpid2` is supported properly on Windows.  We currently use the `win32-process` gem which monkey patches some of the built-in Ruby process APIs ... the gem used to have its own `waitpid`, but that was removed in 0.7.0 in 2012 (https://github.com/djberg96/win32-process/blob/4babb9e410fe68949bc03ee262aa67927637dd23/CHANGES#L53)

FWIW, Ruby does implement it everywhere as far as I can tell - https://github.com/ruby/ruby/blob/8dd9c12c585375cc9d89441d7518a7846044deb0/process.c#L1024-L1030 - but I'm always a little leary of the Ruby Process helpers.